tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1280 @@
+    /// it supports providing security attributes.
+    ///
+    /// [new]: NamedPipeClientOptions::new
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `attrs` points to an initialized instance
+    /// of a [SECURITY_ATTRIBUTES] structure.
+    ///
+    /// [SECURITY_ATTRIBUTES]: [crate::winapi::um::minwinbase::SECURITY_ATTRIBUTES]
+    pub unsafe fn create_with_security_attributes(
+        &self,
+        addr: impl AsRef<OsStr>,
+        attrs: *mut (),
+    ) -> io::Result<NamedPipe> {
+        let addr = encode_addr(addr);
+
+        // NB: We could use a platform specialized `OpenOptions` here, but since
+        // we have access to winapi it ultimately doesn't hurt to use
+        // `CreateFile` explicitly since it allows the use of our already
+        // well-structured wide `addr` to pass into CreateFileW.
+        let h = fileapi::CreateFileW(
+            addr.as_ptr(),
+            self.desired_access,
+            0,
+            attrs as *mut _,
+            fileapi::OPEN_EXISTING,
+            winbase::FILE_FLAG_OVERLAPPED | winbase::SECURITY_IDENTIFICATION,
+            ptr::null_mut(),
+        );
+
+        if h == handleapi::INVALID_HANDLE_VALUE {
+            return Err(io::Error::last_os_error());
+        }
+
+        let io = mio_windows::NamedPipe::from_raw_handle(h);
+        let io = PollEvented::new(io)?;
+
+        Ok(NamedPipe { io })
+    }
+}
+
+/// The pipe mode of a [NamedPipe].
+///
+/// Set through [NamedPipeOptions::pipe_mode].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[non_exhaustive]
+pub enum PipeMode {
+    /// Data is written to the pipe as a stream of bytes. The pipe does not
+    /// distinguish bytes written during different write operations.
+    ///
+    /// Corresponds to [PIPE_TYPE_BYTE][crate::winapi::um::winbase::PIPE_TYPE_BYTE].
+    Byte,
+    /// Data is written to the pipe as a stream of messages. The pipe treats the
+    /// bytes written during each write operation as a message unit. Any reading
+    /// function on [NamedPipe] returns [ERROR_MORE_DATA] when a message is not
+    /// read completely.
+    ///
+    /// Corresponds to [PIPE_TYPE_MESSAGE][crate::winapi::um::winbase::PIPE_TYPE_MESSAGE].
+    ///
+    /// [ERROR_MORE_DATA]: crate::winapi::shared::winerror::ERROR_MORE_DATA
+    Message,
+}
+
+/// Indicates the end of a [NamedPipe].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[non_exhaustive]
+pub enum PipeEnd {
+    /// The [NamedPipe] refers to the client end of a named pipe instance.
+    ///
+    /// Corresponds to [PIPE_CLIENT_END][crate::winapi::um::winbase::PIPE_CLIENT_END].
+    Client,
+    /// The [NamedPipe] refers to the server end of a named pipe instance.
+    ///
+    /// Corresponds to [PIPE_SERVER_END][crate::winapi::um::winbase::PIPE_SERVER_END].
+    Server,
+}
+
+/// Information about a named pipe.
+///
+/// Constructed through [NamedPipe::info].
+#[derive(Debug)]
+pub struct PipeInfo {
```

Comment:
I assume this type is fixed (won't change) because of the winapi?

Reply:
Unlikely to change without a new fn from Microsoft. But it doesn't hurt to mark it as `#[non_exhaustive]` as well.